Issue 5442 - Search results are different between RHDS10 and RHDS11

mreynolds389 opened the pull request:
Bug Description:  

In 1.4.x we introduced a change that was overly strict about how a search on a non-existent subtree returned its error code.  It was changed from returning an error 32 to an error 0 with  zero entries returned.

Fix Description:  

Change the server behavior back to returning error 32 in this case.

relates: https://github.com/389ds/389-ds-base/issues/4542

mreynolds389 commented on ldap/servers/plugins/acl/acl.c:
```diff
@@ -2773,64 +2743,35 @@ acl__resource_match_aci(Acl_PBlock *aclpb, aci_t *aci, int skip_attrEval, int *a
-            } else {
-                matches = ACL_FALSE;
-            }
         }
```
This code removal above is the "fix"...

mreynolds389 replied:
The removal of this code does not cause any regressions in the ACI test suite, but I am not sure it is 100% valid either because the purpose of this code block is not entirely clear.   Please review...

Firstyear replied:
Doing some git investigation (and having to checkout from before the great code format clean up of 2017), I can see the following in git blame:

```
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2790)    } else if ( (c_attrEval) ||
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2791)                (aci->aci_type & ACI_TARGET_ATTR)) {
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2792)            if ((aci_right & ACL_RIGHTS_TARGETATTR_NOT_NEEDED) &&
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2793)                (aclpb->aclpb_access & ACL_RIGHTS_TARGETATTR_NOT_NEEDED)) {
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2794)                    /*
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2795)                    ** Targetattr rule doesn't  make any sense
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2796)                    ** in this case. So select this rule
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2797)                    ** default: matches = ACL_TRUE;
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2798)                    */
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2799)                    ;
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2800)            } else if (aci_right & SLAPI_ACL_WRITE &&
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2801)                      (aci->aci_type & ACI_TARGET_ATTR) &&
fe0491c49f (Ludwig Krispenz          2013-07-09 16:11:12 +0200 2802)                      !(c_attrEval) &&
fe0491c49f (Ludwig Krispenz          2013-07-09 16:11:12 +0200 2803)                      (aci->aci_type & ACI_HAS_ALLOW_RULE)) {
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2804)                    /* We need to handle modrdn operation.  Modrdn doesn't
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2805)                    ** change any attrs but changes the RDN and so (attr=NULL).
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2806)                    ** Here we found an acl which has a targetattr but
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2807)                    ** the resource doesn't need one. In that case, we should
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2808)                    ** consider this acl.
fe0491c49f (Ludwig Krispenz          2013-07-09 16:11:12 +0200 2809)                    ** the opposite is true if it is a deny rule, only a deny without
fe0491c49f (Ludwig Krispenz          2013-07-09 16:11:12 +0200 2810)                    ** any targetattr should deny modrdn
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2811)                    ** default: matches = ACL_TRUE;
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2812)                    */
^b2093e301 (cvsadm                   2005-01-21 00:44:34 +0000 2813)                    ;

```

It looks like the majority of this code is "quite old" to say the least, but what is interesting is the condition change from ludwig in 2013. According to git its:

```
commit fe0491c49fdb180f1bd291ba78148da7c58b5e90
Author: Ludwig Krispenz <lkrispen@redhat.com>
Date:   Tue Jul 9 16:11:12 2013 +0200

    Ticket 47399 - RHDS denies MODRDN access if ACI list contains any DENY rule

    Bug Description:  if there is a deny rule targeting a specific attribute it also
                      denies the modrdn operation

    Fix Description:   only apply deny rules to modrdn if no target attr exists

    https://fedorahosted.org/389/ticket/47339

    Reviewed by: Rich, thanks
```

So I'm not sure the delete of this block is the right change here, especialy given how ... fun the aci code is at the best of times. I think we need some more research?

mreynolds389 committed:
Issue 5442 - Search results are different between RHDS10 and RHDS11

Bug Description:  In 1.4.x we introduced a change that was overly strict about
                  how a search on a non-existent subtree returned its error code.
                  It was changed from returning an error 32 to an error 0 with
                  zero entries returned.

Fix Description:  When finding the entry and processing acl's make sure to
                  gather the aci's that match the resource even if the resource
                  does not exist.  This requires some extra checks when processing
                  the target attribute.

relates: https://github.com/389ds/389-ds-base/issues/4542

Reviewed by: firstyear, elkris, and tbordaz (Thanks!)

Apply Thierry's changes

round 2

Apply more suggestions from Thierry

tbordaz commented on ldap/servers/slapd/back-ldbm/findentry.c:
```diff
@@ -207,7 +207,11 @@ find_entry_internal_dn(
+                    char *dummyAttrs[2] = {NULL, NULL};
+
+                    dummyAttrs[0] = dummyAttr;
+                    err = plugin_call_acl_plugin(pb, me->ep_entry, dummyAttrs, NULL, acl_type,
```
Instead of calling directly the acl plugin you could also use slapi_access_allowed (like https://github.com/389ds/389-ds-base/blob/master/ldap/servers/plugins/sync/sync_refresh.c#L40)

tbordaz commented on dirsrvtests/tests/suites/acl/misc_test.py:
```diff
@@ -408,14 +408,94 @@ def test_do_bind_as_201_distinct_users(topo, clean, aci_of_user):
+        conn = UserAccount(topo.standalone, USER_DN).bind(PW_DM)
+        subtree = Domain(conn, "ou=does_not_exist," + DEFAULT_SUFFIX)
+        subtree.get_attr_vals_utf8_l('objectclass')
+
```
It should also return NO_SUCH_OBJECT for base search and one-level search.
Also is it possible to test something like: 

subtree = Domain(conn, "ou=even_more,ou=does_not_exist," + DEFAULT_SUFFIX)

tbordaz reviewed (commented):
Except those minor comments the fix looks good to me
